File: mvvm/JetLiYu/app/src/main/java/pri/weiqiang/jet/liyujapanese/ui/word/WordFragmentViewModel.java
package pri.weiqiang.jet.liyujapanese.ui.word;

import android.text.TextUtils;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;
import pri.weiqiang.jet.liyujapanese.App;
import pri.weiqiang.jet.liyujapanese.data.database.VocabDatabase;
import pri.weiqiang.jet.liyujapanese.data.database.Word;

public class WordFragmentViewModel extends ViewModel {

    private String TAG = WordFragmentViewModel.class.getSimpleName();
    private LiveData<List<Word>> wordList;
    private String initLessonid = "新标日初级_01";
    private static final String QUERY_KEY = "LessonId";
    private final SavedStateHandle mSavedStateHandler;

    public LiveData<List<Word>> getWordList() {
        return wordList;
    }

    public WordFragmentViewModel(SavedStateHandle savedStateHandle) {
        mSavedStateHandler = savedStateHandle;
        //无法直接使用update(String lessonId),不会响应改变，因为wordList的地址已经改变，不是原来的wordList了，需要重新增加观察者
        wordList = Transformations.switchMap(
                savedStateHandle.getLiveData(QUERY_KEY, "新标日初级_01"),
                lessonId -> {
                    if (TextUtils.isEmpty(lessonId)) {
                        //database中直接使用App.getInstance不要从ViewModel传入任何Context
                        return VocabDatabase.getInstance(App.getInstance()).wordDao().getWordByLessonId(lessonId);
                    }
                    return VocabDatabase.getInstance(App.getInstance()).wordDao().getWordByLessonId(lessonId);
                });
    }

    public void setQuery(String query) {
        mSavedStateHandler.set(QUERY_KEY, query);
    }
    //wrong:不会被观察者观察到
//    public void update(String lessonId){
//        wordList = VocabDatabase.getInstance(App.getInstance()).wordDao().getWordByLessonId(lessonId);
//    }
}

File: mvvm/KtJetLiYu/app/src/main/java/pri/weiqiang/jet/kt/liyujapanese/ui/viewmodel/WordFragmentViewModel.kt
package pri.weiqiang.jet.kt.liyujapanese.ui.viewmodel

import android.text.TextUtils
import androidx.lifecycle.LiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import pri.weiqiang.jet.kt.liyujapanese.App
import pri.weiqiang.jet.kt.liyujapanese.data.bean.Word
import pri.weiqiang.jet.kt.liyujapanese.data.database.VocabDatabase

class WordFragmentViewModel(private val mSavedStateHandler: SavedStateHandle) : ViewModel() {
    private val TAG = WordFragmentViewModel::class.java.simpleName
    val wordList: LiveData<List<Word>>
    private val initLessonId = "新标日初级_01"

    fun setQuery(query: String) {
        mSavedStateHandler.set(QUERY_KEY, query)
    }

    companion object {
        private const val QUERY_KEY = "LessonId"
    }

    init {
        wordList = Transformations.switchMap(
            mSavedStateHandler.getLiveData(
                QUERY_KEY,
                "新标日初级_01"
            )
        ) { lessonId: String ->
            if (TextUtils.isEmpty(lessonId)) {
                return@switchMap VocabDatabase.getInstance(App.instance())?.wordDao()
                    ?.getWordByLessonId(lessonId)
            }
            VocabDatabase.getInstance(App.instance())?.wordDao()?.getWordByLessonId(lessonId)
        }
    }
}